In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import SGDRegressor
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.metrics import r2_score
import numpy as np

# Assignment 2 - Regression
### Predict the TOTAL COMPENSATION for this year.

The data file provided is a salary survey for tech workers in Europe. We want to predict the total amount of compensation they bring in each year, based off of the details of their work.

Some notes that will be important:
<ul>
    <li>The total compensation will need to be constructed, there is a column for salary, "Yearly brutto salary (without bonus and stocks) in EUR", as well as a column for bonus compensation, "Yearly bonus + stocks in EUR".
    <li>Some categorical variables will need some work, and there isn't generally an exact answer. The main concern is things with categories that have a bunch of values with a very small count. For example, if there is only 1 person in City X, then that value likely needs to be addressed. We don't want it encoded into a new column of one 1 and thousands of 0s.
    <li>There is an article exploring some of the data here: https://www.asdcode.de/2021/01/it-salary-survey-december-2020.html
    <li>Imputation and a bit of data manipulation will be required.
    <li>Use any regression method you'd like. Some ones are closely related to what we've done, you may want to look at them, e.g. ExtraTreesRegressor.
    <li>Initial accurracy, and potentially final accuracy, may not be great. When I made a plain model will little optimization the errors were large and the R2 was low. There is lots of room for optimization.
    <li>Research challenge - try some work on the target, look into TransformedTargetRegressor and see if that helps. Recall in stats when we had skewed distributions... Maybe it helps, maybe it doesn't.
    <li>EDA and data prep are up to you - you'll probably need to do a little exploring to figure out what cleanup is needed. When I did it, I did things kind of iteratively when I did it. For example, look at the value counts, figure out how to treat the different categories, clean something up, look at the results, potentially repeat if needed. After you figure out what needs to be done, you may be able to take some of those steps and incorporate them into a pipeline to be cleaner....
    <li><b>CRITICAL - Please make sure your code runs with RUN ALL. It should load the data that you're given, do all the processing, and spit out results. Comment out or remove anything that you've cleaned up and don't need - e.g. if you scaled a value manually, then moved that into a pipeline, don't leave the original scaling code active when the file is run.</b>
</ul>

### Details and Deliverables

You'll need to build code to produce the predictions. In particular, there's a few things that'll be marked:
<ul>
    <li>Please add a markdown cell at the bottom, and put in a few notes addressing the following:
    <ul>
        <li> Accuracy of your models with/without feature selection. Include both train/test for each. Please use R2 and RMSE.
        <li> Feature Selection - Please identify what you did for feature selection. No need for a long explaination, something along the lines of "I did X, and the result was that 4 features were removed". Try at least 2 things.
        <li> Hyperparameter Changes / Grid Search Improvements. What did you try, and why. Similar explaination to above, short.
        <li> Overall this section should be roughly as long as this intro block - just outline what the results were, what you did to improve, and the results after.
        <li> If you could use titles/bullet points I'd really appreciate it.
    </ul>
    <li>Grade Breakdown:
    <ul>
        <li> Code is readable, there are comments: 20%
        <li> Explaination as defined above: 60% (20% each point)
        <li> Accuracy: 20% As compared to everyone else. This will be generously graded, I won't be surprised if overall accuracy is low for most people.
    </ul>
</ul>

In [2]:
#Load Data
df = pd.read_csv("Euro_Salary.csv")
df.describe(include="all").T
# First, I load the salary survey dataset and take a quick look at it.
# This helps me understand what columns exist and what kind of cleaning will be needed.


count unique  \
Timestamp                                             1253   1248   
Age                                                 1226.0    NaN   
Gender                                                1243      3   
City                                                  1253    119   
Position                                              1247    148   
Total years of experience                             1237     48   
Years of experience in Germany                        1221     53   
Seniority level                                       1241     24   
Your main technology / programming language           1126    256   
Other technologies/programming languages you us...    1096    562   
Yearly brutto salary (without bonus and stocks)...  1253.0    NaN   
Yearly bonus + stocks in EUR                           829    168   
Number of vacation days                               1185     45   
Employment status                                     1236     11   
Сontract duration                                     1224      3   
Main language at work                                 1237     14   
Company size                                          1235      5   
Company type                                          1228     63   

                                                                        top  \
Timestamp                                               24/11/2020 13:55:19   
Age                                                                     NaN   
Gender                                                                 Male   
City                                                                 Berlin   
Position                                                  Software Engineer   
Total years of experience                                                10   
Years of experience in Germany                                            2   
Seniority level                                                      Senior   
Your main technology / programming language                            Java   
Other technologies/programming languages you us...  Javascript / Typescript   
Yearly brutto salary (without bonus and stocks)...                      NaN   
Yearly bonus + stocks in EUR                                              0   
Number of vacation days                                                  30   
Employment status                                        Full-time employee   
Сontract duration                                        Unlimited contract   
Main language at work                                               English   
Company size                                                          1000+   
Company type                                                        Product   

                                                    freq             mean  \
Timestamp                                              2              NaN   
Age                                                  NaN        32.509788   
Gender                                              1049              NaN   
City                                                 681              NaN   
Position                                             387              NaN   
Total years of experience                            138              NaN   
Years of experience in Germany                       195              NaN   
Seniority level                                      565              NaN   
Your main technology / programming language          184              NaN   
Other technologies/programming languages you us...    44              NaN   
Yearly brutto salary (without bonus and stocks)...   NaN  80279042.578715   
Yearly bonus + stocks in EUR                         227              NaN   
Number of vacation days                              488              NaN   
Employment status                                   1190              NaN   
Сontract duration                                   1159              NaN   
Main la

In [4]:
df.head()

Timestamp   Age Gender    City           Position   \
0  24/11/2020 11:14:15  26.0   Male  Munich   Software Engineer   
1  24/11/2020 11:14:16  26.0   Male  Berlin   Backend Developer   
2  24/11/2020 11:14:21  29.0   Male  Berlin   Software Engineer   
3  24/11/2020 11:15:24  28.0   Male  Berlin  Frontend Developer   
4  24/11/2020 11:15:46  37.0   Male  Berlin   Backend Developer   

  Total years of experience Years of experience in Germany Seniority level  \
0                         5                              3          Senior   
1                         7                              4          Senior   
2                        12                              6            Lead   
3                         4                              1          Junior   
4                        17                              6          Senior   

  Your main technology / programming language  \
0                                  TypeScript   
1                                        Ruby   
2                     Javascript / Typescript   
3                                  Javascript   
4                                     C# .NET   

  Other technologies/programming languages you use often  \
0                    Kotlin, Javascript / Typescript       
1                                                NaN       
2                    Javascript / Typescript, Docker       
3                                                NaN       
4                             .NET, SQL, AWS, Docker       

   Yearly brutto salary (without bonus and stocks) in EUR  \
0                                            80000.0        
1                                            80000.0        
2                                           120000.0        
3                                            54000.0        
4                                            62000.0        

  Yearly bonus + stocks in EUR Number of vacation days  \
0                         5000                      30   
1                          NaN                      28   
2                       120000                      30   
3                          NaN                      24   
4                          NaN                      29   

            Employment status   Сontract duration Main language at work  \
0          Full-time employee  Unlimited contract               English   
1          Full-time employee  Unlimited contract               English   
2  Self-employed (freelancer)  Temporary contract               English   
3          Full-time employee  Unlimited contract               English   
4          Full-time employee  Unlimited contract               English   

  Company size Company type  
0       51-100      Product  
1     101-1000      Product  
2     101-1000      Product  
3       51-100      Startup  
4     101-1000      Product

In [5]:
# Build the target we want to predict: Total yearly compensation (EUR)

# The assignment requires predicting total compensation.
# Since base salary and bonus are separate columns, I combine them into one target variable.

salary_col = "Yearly brutto salary (without bonus and stocks) in EUR"
bonus_col  = "Yearly bonus + stocks in EUR"

# Convert salary column to numeric, coercing errors to NaN, then fill NaN with 0
df[salary_col] = pd.to_numeric(df[salary_col], errors='coerce').fillna(0)

# Convert bonus column to numeric, coercing errors to NaN, then fill NaN with 0
df[bonus_col] = pd.to_numeric(df[bonus_col], errors='coerce').fillna(0)

# Total compensation = base salary + bonus/stocks
df["total_comp"] = df[salary_col] + df[bonus_col]

# Quick check of the target distribution
df["total_comp"].describe()

count    1.253000e+03
mean     8.428361e+07
std      2.829182e+09
min      1.000100e+04
25%      6.000000e+04
50%      7.400000e+04
75%      9.100000e+04
max      1.000000e+11
Name: total_comp, dtype: float64

In [6]:
# Survey data has fake entries like 9,999,999,999. Remove unrealistic totals.
# Keep compensation between 1,000 and 500,000 EUR (simple, reasonable bounds)

# Because this is survey data, some responses contain unrealistic salary values.
# I remove extreme outliers to prevent them from distorting the model.


df = df[(df["total_comp"] >= 1000) & (df["total_comp"] <= 500000)]

df["total_comp"].describe()



count      1248.000000
mean      83717.752869
std       42329.673843
min       10001.000000
25%       60000.000000
50%       73450.000000
75%       91000.000000
max      400000.000000
Name: total_comp, dtype: float64

In [7]:
# Some categorical columns have many unique values.
# To reduce noise and avoid too many one-hot encoded columns,
# I group rare categories into "Other".

rare_threshold = 10

high_card_cols = [
    "City",
    "Your main technology / programming language",
    "Other technologies/programming languages you use often"
]

for c in high_card_cols:
    if c in df.columns:
        vc = df[c].value_counts(dropna=False)
        rare = vc[vc < rare_threshold].index
        df[c] = df[c].where(~df[c].isin(rare), "Other")



In [8]:
# I now separate predictors (X) and the target (y).
# All salary and compensation-related columns are removed to prevent data leakage.

drop_leak_cols = [
    "total_comp",
    "TotalCompEUR",  # <-- common leak if it exists
    salary_col,
    bonus_col,
    "Timestamp"
]

X = df.drop(columns=drop_leak_cols, errors="ignore")
y = df["total_comp"]

# Sanity checks (these should be False)
print("total_comp in X?", "total_comp" in X.columns)
print("TotalCompEUR in X?", "TotalCompEUR" in X.columns)
print("salary_col in X?", salary_col in X.columns)
print("bonus_col in X?", bonus_col in X.columns)

X.head()



total_comp in X? False
TotalCompEUR in X? False
salary_col in X? False
bonus_col in X? False


Age Gender    City           Position  Total years of experience  \
0  26.0   Male  Munich   Software Engineer                         5   
1  26.0   Male  Berlin   Backend Developer                         7   
2  29.0   Male  Berlin   Software Engineer                        12   
3  28.0   Male  Berlin  Frontend Developer                         4   
4  37.0   Male  Berlin   Backend Developer                        17   

  Years of experience in Germany Seniority level  \
0                              3          Senior   
1                              4          Senior   
2                              6            Lead   
3                              1          Junior   
4                              6          Senior   

  Your main technology / programming language  \
0                                  TypeScript   
1                                        Ruby   
2                                       Other   
3                                  Javascript   
4                                       Other   

  Other technologies/programming languages you use often  \
0                                              Other       
1                                                NaN       
2                                              Other       
3                                                NaN       
4                                              Other       

  Number of vacation days           Employment status   Сontract duration  \
0                      30          Full-time employee  Unlimited contract   
1                      28          Full-time employee  Unlimited contract   
2                      30  Self-employed (freelancer)  Temporary contract   
3                      24          Full-time employee  Unlimited contract   
4                      29          Full-time employee  Unlimited contract   

  Main language at work Company size Company type  
0               English       51-100      Product  
1               English     101-1000      Product  
2               English     101-1000      Product  
3               English       51-100      Startup  
4               English     101-1000      Product

In [10]:
# I build a preprocessing and modeling pipeline.
# Numeric features are scaled, and categorical features are one-hot encoded.
# The model is trained and evaluated using R2 and RMSE.


num_cols = X.select_dtypes(include="number").columns.tolist()
cat_cols = [c for c in X.columns if c not in num_cols]

preprocess = ColumnTransformer(
    transformers=[
        ("num", Pipeline([
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler())
        ]), num_cols),

        ("cat", Pipeline([
            ("imputer", SimpleImputer(strategy="most_frequent")),
            ("onehot", OneHotEncoder(handle_unknown="ignore"))
        ]), cat_cols),
    ],
    remainder="drop"
)

pipe = Pipeline([
    ("prep", preprocess),
    ("model", Ridge(alpha=1.0))
])

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

pipe.fit(X_train, y_train)

pred_train = pipe.predict(X_train)
pred_test  = pipe.predict(X_test)

print("TRAIN R2:", r2_score(y_train, pred_train))
print("TRAIN RMSE:", np.sqrt(mean_squared_error(y_train, pred_train)))
print("TEST  R2:", r2_score(y_test, pred_test))
print("TEST  RMSE:", np.sqrt(mean_squared_error(y_test, pred_test)))

TRAIN R2: 0.547673477951679
TRAIN RMSE: 28134.24989359479
TEST  R2: 0.14006635845616366
TEST  RMSE: 40962.402677362545


In [11]:
# As a feature selection experiment, I drop a high-cardinality technology column
# and compare model performance to see if it improves generalization.

drop_col = "Other technologies/programming languages you use often"

X_fs = X.drop(columns=[drop_col], errors="ignore")

# Redefine num_cols and cat_cols based on X_fs
num_cols_fs = X_fs.select_dtypes(include="number").columns.tolist()
cat_cols_fs = [c for c in X_fs.columns if c not in num_cols_fs]

# Create a new preprocess transformer for X_fs
preprocess_fs = ColumnTransformer(
    transformers=[
        ("num", Pipeline([
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler())
        ]), num_cols_fs),

        ("cat", Pipeline([
            ("imputer", SimpleImputer(strategy="most_frequent")),
            ("onehot", OneHotEncoder(handle_unknown="ignore"))
        ]), cat_cols_fs),
    ],
    remainder="drop"
)

# Create a new pipeline for X_fs
pipe_fs = Pipeline([
    ("prep", preprocess_fs),
    ("model", Ridge(alpha=1.0))
])

X_train_fs, X_test_fs, y_train_fs, y_test_fs = train_test_split(
    X_fs, y, test_size=0.2, random_state=42
)

pipe_fs.fit(X_train_fs, y_train_fs)
pred_fs = pipe_fs.predict(X_test_fs)

print("AFTER DROPPING:", drop_col)
print("TEST R2:", r2_score(y_test_fs, pred_fs))
print("TEST RMSE:", np.sqrt(mean_squared_error(y_test_fs, pred_fs)))

AFTER DROPPING: Other technologies/programming languages you use often
TEST R2: 0.13510354317812245
TEST RMSE: 41080.432934695535


In [12]:
# To improve model performance, I tune the Ridge regularization parameter (alpha)
# using GridSearchCV with cross-validation.


param_grid = {"model__alpha": [0.1, 1, 10, 100]}

grid = GridSearchCV(pipe, param_grid, cv=5, scoring="r2")
grid.fit(X_train, y_train)

print("Best alpha:", grid.best_params_)
print("Best CV R2:", grid.best_score_)

best_model = grid.best_estimator_
best_pred = best_model.predict(X_test)

print("TUNED TEST R2:", r2_score(y_test, best_pred))
print("TUNED TEST RMSE:", np.sqrt(mean_squared_error(y_test, best_pred)))

Best alpha: {'model__alpha': 10}
Best CV R2: 0.1739030060057239
TUNED TEST R2: 0.20467106224485276
TUNED TEST RMSE: 39393.661160918826


# Answers and Explainations
(Expand/modify as needed)

### Results
The baseline Ridge regression model achieved a training R² of 0.55 and a test R² of 0.14. The training RMSE was approximately €28,134, and the test RMSE was approximately €40,962. The drop from training to test indicates some overfitting, but the model still captures meaningful patterns related to experience, role, and technology.

After feature selection and hyperparameter tuning, performance was compared to the baseline to evaluate improvements in generalization.
### Feature Selection Activities
Removed all salary, bonus, and compensation-related columns from the predictors to prevent data leakage.

Grouped rare categorical values (e.g., City and technology columns with fewer than 10 observations) into “Other” to reduce sparse one-hot encoding and improve stability.

Dropped the high-cardinality column “Other technologies/programming languages you use often” and compared performance to assess its impact.
### Hyperparameter Changes
Used GridSearchCV to tune the Ridge regression regularization parameter (alpha) over values [0.1, 1, 10, 100] using 5-fold cross-validation. The best alpha was selected based on cross-validation R², and the tuned model was evaluated on the test set to compare performance against the baseline.